File: 0567-permutation-in-string/0567-permutation-in-string.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        # approach: sliding window, two hashmaps of char : frequency
        
        if len(s1) > len(s2):
            return False
        
        d1, d2 = [0] * 26, [0] * 26
        
        for c in s1:
            d1[ord(c) - ord('a')] += 1

        l = 0
        for r in range(len(s2)):
            index = ord(s2[r]) - ord('a')
            d2[index] += 1
            
            if d1[index] == 0 or d2[index] > d1[index]:
                while d1[index] != d2[index] and l <= r:
                    d2[ord(s2[l]) - ord('a')] -= 1
                    l += 1
            
            if d1 == d2:
                return True
        
        return False

File: 0567-permutation-in-string/0567-permutation-in-string.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        // approach: two arrays to store char counts
        
        if (s1.length() > s2.length()) {
            return false;
        }
        
        Map<Character, Integer> s1Map = new HashMap<>();
        Map<Character, Integer> s2Map = new HashMap<>();
        
        for (char c = 'a'; c <= 'z'; c++) {
            s1Map.put(c, 0);
            s2Map.put(c, 0);
        }
        
        for (int i = 0; i < s1.length(); i++) {
            char c1 = s1.charAt(i);
            char c2 = s2.charAt(i);
            
            s1Map.put(c1, s1Map.get(c1) + 1);
            s2Map.put(c2, s2Map.get(c2) + 1);
        }
        
        if (s1Map.equals(s2Map)) {
            return true;
        }
        
        int l = 0;
        int r = s1.length() - 1;
        
        while (r < s2.length()) {
            char lChar = s2.charAt(l);
            s2Map.put(lChar, s2Map.get(lChar) - 1);
            
            l++;
            r++;
            
            if (r < s2.length()) {
                char rChar = s2.charAt(r);
                s2Map.put(rChar, s2Map.get(rChar) + 1);
            }
            
            if (s1Map.equals(s2Map)) {
                return true;
            }
        }
        
        return false;
    }
}